### PR TITLE
Fix BE can't be grayscale upgraded

### PR DESCRIPTION
**Pull request by chenhao7253886**
Support BE grayscale upgrade.

### CODE REVIEW

File: be/src/runtime/lib_cache.h [hunk 1]
```diff
@@ -161,6 +161,36 @@ class LibCache {
...
         boost::unique_lock<boost::mutex>* entry_lock, LibCacheEntry** entry);
 
+    // map "palo" to "doris" in symbol, only for grayscale upgrading
+    std::string get_real_symbol(const std::string& symbol) {
+        const std::string& str1 = replace_all(symbol, "8palo_udf", "9doris_udf");
+        const std::string& str2 = replace_all(str1, "4palo", "5doris"); 
+        return str2;
+    }
+
+    std::string replace_all(const std::string& str, const std::string& find, const std::string& replace) {
```
Comment (imay) [inline be/src/runtime/lib_cache.h]: you can use c++11 std::regex_replace

File: be/src/runtime/lib_cache.h [hunk 2]
```diff
@@ -161,6 +161,36 @@ class LibCache {
         const std::string& hdfs_lib_file, LibType type,
         boost::unique_lock<boost::mutex>* entry_lock, LibCacheEntry** entry);
 
+    // map "palo" to "doris" in symbol, only for grayscale upgrading
+    std::string get_real_symbol(const std::string& symbol) {
+        const std::string& str1 = replace_all(symbol, "8palo_udf", "9doris_udf");
+        const std::string& str2 = replace_all(str1, "4palo", "5doris");
```
Comment (imay) [inline be/src/runtime/lib_cache.h]: you can't use reference here, just use std::string, complier will do RVO

File: be/src/runtime/lib_cache.cpp [hunk 3]
```diff
@@ -126,10 +126,11 @@ LibCache::LibCacheEntry::~LibCacheEntry() {
 Status LibCache::get_so_function_ptr(
         const std::string& hdfs_lib_file, const std::string& symbol,
         void** fn_ptr, LibCacheEntry** ent, bool quiet) {
+    const std::string& real_symbol = get_real_symbol(symbol);
```
Comment (imay) [inline be/src/runtime/lib_cache.cpp]: you can change argument symbol to origin_symbol, then you needn't change following symbol to real_symbol

**Review by imay — CHANGES_REQUESTED** [review]

File: be/src/runtime/lib_cache.h (hunk 1 above)
Comment (chenhao7253886) [reply be/src/runtime/lib_cache.h]: ok, i will replace it.

File: be/src/runtime/lib_cache.h (hunk 2 above)
Comment (chenhao7253886) [reply be/src/runtime/lib_cache.h]: oh, variables in the stack will be released , when function finished. I forget , i will correct it.

File: be/src/runtime/lib_cache.cpp (hunk 3 above)
Comment (chenhao7253886) [reply be/src/runtime/lib_cache.cpp]: ok, it is more simple。i will change it.

**Review by imay — CHANGES_REQUESTED** [review]

**Review by imay — APPROVED** [review]
LGTM

**Comment by kangkaisen** [comment]
BE can't be grayscale upgraded even if I using the latest code.

I upgraded one BE from 0.8.2.1 to latest version.

When I query,  mysql client error is following;

```
ERROR 1064 (HY000): Ocurrs time out with specfied time xxx MILLISECONDS
or
ERROR 1064 (HY000): failed to send batch
```
The latest BE warning logs:
```
W1109 14:43:35.397363 28981 data_stream_sender.cpp:125] failed to send brpc batch, error=No such method, error_text=[E1002][10.24.
104.124:8060][E1002]Fail to find method=doris.PInternalService/transmit_data
```
The 0.8.2.1 BE warning logs
```
W1109 14:43:37.406872 10384 data_stream_sender.cpp:149] failed to send brpc batch, error=No such method, error_text=[E1002][10.20.93.37:8060][E1002]Fail to find method=palo.PInternalService/transmit_data
```

The cause for this error is obvious， So I think this issue have not  been fixed completely.

**Comment by kangkaisen** [comment]
OK.

**Pull request merged**